File: internal/index/tokenizer.go
package index

import (
	"strings"
	"unicode"
)

// set of english words excluded from indexing
var stopWords = map[string]bool{
	"a": true, "an": true, "and": true, "are": true, "as": true,
	"at": true, "be": true, "been": true, "but": true, "by": true,
	"do": true, "for": true, "from": true, "get": true, "has": true,
	"have": true, "he": true, "her": true, "his": true, "how": true,
	"i": true, "if": true, "in": true, "is": true, "it": true,
	"its": true, "just": true, "me": true, "my": true, "no": true,
	"not": true, "of": true, "on": true, "or": true, "our": true,
	"out": true, "she": true, "so": true, "than": true, "that": true,
	"the": true, "their": true, "them": true, "then": true, "there": true,
	"they": true, "this": true, "to": true, "up": true, "us": true,
	"was": true, "we": true, "were": true, "what": true, "when": true,
	"which": true, "who": true, "will": true, "with": true, "you": true,
	"your": true,
}

// convert text into a normalised slice of tokens suitable for indexing
func Tokenize(text string, minLength int) []string {
	// replace punctuation with spaces 
	var b strings.Builder
	for _, r := range text {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(unicode.ToLower(r))
		} else {
			b.WriteRune(' ')
		}
	}

	raw := strings.Fields(b.String())
	out := make([]string, 0, len(raw))
	for _, tok := range raw {
		if len(tok) < minLength {
			continue
		}
		if stopWords[tok] {
			continue
		}
		out = append(out, tok)
	}
	return out
}

// returns a map of token to count within text
// used at indexing time to determine term frequency
func TermFrequencies(tokens []string) map[string]int {
	freq := make(map[string]int, len(tokens))
	for _, t := range tokens {
		freq[t]++
	}
	return freq
}

File: app/index/tokenizer.py
"""
Tokenizer - direct translation of internal/index/tokenizer.go.
 
The logic is identical: lowercase, replace non-alphanumeric characters with
spaces, split, skip stop words and short tokens. The only Python-specific
note is that we use str.isalnum() and str.lower() instead of Go's
unicode.IsLetter/unicode.IsDigit/unicode.ToLower, but they're equivalent.
"""
 
from __future__ import annotations
 
STOP_WORDS = frozenset({
    "a", "an", "and", "are", "as", "at", "be", "been", "but", "by",
    "do", "for", "from", "get", "has", "have", "he", "her", "his", "how",
    "i", "if", "in", "is", "it", "its", "just", "me", "my", "no",
    "not", "of", "on", "or", "our", "out", "she", "so", "than", "that",
    "the", "their", "them", "then", "there", "they", "this", "to", "up",
    "us", "was", "we", "were", "what", "when", "which", "who", "will",
    "with", "you", "your",
})
 
 
def tokenize(text: str, min_length: int) -> list[str]:
    """
    Lowercase, strip punctuation, remove stop words and short tokens.
    Identical output to the Go version's Tokenize() for the same input.
    """
    # Replace non-alphanumeric characters with spaces so "hello,world"
    # becomes "hello world" rather than "helloworld".
    cleaned = "".join(c.lower() if c.isalnum() else " " for c in text)
 
    return [
        tok
        for tok in cleaned.split()
        if len(tok) >= min_length and tok not in STOP_WORDS
    ]
 
 
def term_frequencies(tokens: list[str]) -> dict[str, int]:
    """Return a mapping of token > count. Used at index time."""
    freq: dict[str, int] = {}
    for tok in tokens:
        freq[tok] = freq.get(tok, 0) + 1
    return freq
 
